src/util/metadata/side_metadata/global.rs

```diff
@@ -157,19 +203,19 @@ impl SideMetadataSpec {
         #[cfg(feature = "extreme_assertions")]
         let _lock = sanity::SANITY_LOCK.lock().unwrap();
 
-        // yiluowei: Not Sure but this assertion seems too strict for Immix recycled lines
-        #[cfg(not(feature = "global_alloc_bit"))]
-        debug_assert!(start.is_aligned_to(BYTES_IN_PAGE) && meta_byte_lshift(self, start) == 0);
-
         #[cfg(feature = "extreme_assertions")]
         sanity::verify_bzero(self, start, size);
 
-        let meta_start = address_to_meta_address(self, start);
+        let zero = |data_start: Address, data_end: Address| {
+            let meta_start = address_to_meta_address(self, data_start);
+            let meta_start_shift = meta_byte_lshift(self, data_start);
+            let meta_end = address_to_meta_address(self, data_end);
```

Comment:
It is OK if the metadata is contiguous.  But if it is discontiguous, then `meta_start` and `meta_end` may be in two unrelated chunks.

Reply:
I changed `zero` to `zero_contiguous` which uses the same implementation as what you reviewed, and added `zero_discontiguous` which computes the total bits that need to be zeroed without computing the end address.